src/box/lua/call.c

```diff
@@ -92,61 +95,239 @@ get_call_serializer(void)
+{
+	lexer->src = src;
+	lexer->src_len = src_len;
+	lexer->offset = 0;
+	lexer->symbol_count = 0;
+}
+
+/* Check if lexer has finished parsing. */
+static bool
+lexer_is_eof(const struct lexer *lexer)
+{
+	return lexer->offset == lexer->src_len;
+}
+
+/* Read a single symbol from a string starting from an offset. */
 static int
-box_lua_find(lua_State *L, const char *name, const char *name_end)
+lexer_read_symbol(struct lexer *lexer, UChar32 *out)
 {
-	int index = LUA_GLOBALSINDEX;
-	int objstack = 0, top = lua_gettop(L);
-	const char *start = name, *end;
-
-	while ((end = (const char *) memchr(start, '.', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! lua_istable(L, -1)) {
-			diag_set(ClientError, ER_NO_SUCH_PROC,
-				 name_end - name, name);
-			return -1;
-		}
-		start = end + 1; /* next piece of a.b.c */
-		index = lua_gettop(L); /* top of the stack */
+	if (lexer_is_eof(lexer)) {
+		*out = U_SENTINEL;
+		return -1;
 	}
+	U8_NEXT(lexer->src, lexer->offset, lexer->src_len, *out);
+	if (*out == U_SENTINEL)
+		return -1;
+	++lexer->symbol_count;
+	return 0;
+}
 
-	/* box.something:method */
-	if ((end = (const char *) memchr(start, ':', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! (lua_istable(L, -1) ||
-			lua_islightuserdata(L, -1) || lua_isuserdata(L, -1) )) {
-				diag_set(ClientError, ER_NO_SUCH_PROC,
-					  name_end - name, name);
+/* Rollback one symbol offset. */
+static void
+lexer_revert_symbol(struct lexer *lexer, size_t offset)
+{
+	lexer->offset = offset;
+	--lexer->symbol_count;
+}
+
+/* Fast forward when it is known that a symbol is 1-byte char. */
+static inline void
+lexer_skip_char(struct lexer *lexer)
+{
+	++lexer->offset;
+	++lexer->symbol_count;
+}
+
+/* Get a current symbol as a 1-byte char. */
+static char
+lexer_current_char(const struct lexer *lexer)
+{
+	return lexer->src[lexer->offset];
+}
+
+/*
+ * Parse string identifier in quotes. Lexer stops right after the
+ * closing quote.
+ */
+static inline int
+lexer_parse_string(struct lexer *lexer, const char **str, size_t *str_len,
+		   UChar32 quote)
+{
+	assert(lexer->offset < lexer->src_len);
+	assert(quote == lexer_current_char(lexer));
+	/* The first symbol is always char  - ' or ". */
+	lexer_skip_char(lexer);
+	size_t str_offset = lexer->offset;
+	*str = lexer->src + str_offset;
+	UChar32 c;
+	while (lexer_read_symbol(lexer, &c) == 0) {
+		if (c == quote) {
+			*str_len = lexer->offset - str_offset - 1;
+			if (*str_len == 0)
 				return -1;
+			return 0;
 		}
-
-		start = end + 1; /* next piece of a.b.c */
-		index = lua_gettop(L); /* top of the stack */
-		objstack = index - top;
 	}
+	return -1;
+}
 
+/*
+ * Parse digit sequence into integer until non-digit is met. Lexer stops right
+ * after the last digit.
+ */
+static int
+lexer_parse_integer(struct lexer *lexer, int *num)
```

Comment:
You need to support negative values

```
tarantool> _G.map = {[-1] = print, [0] = print, [1] = print}
---
...

tarantool> c:call('map[0]', {1, 2, 3})
1	2	3
---
...

tarantool> c:call('map[1]', {1, 2, 3})
1	2	3
---
...

tarantool> c:call('map[-1]', {1, 2, 3})
---
- error: Procedure 'map[-1]' is not defined
...
```

Reply:
I think it is possible to think of a lot of exotic cases, but AFAIC, we only want to support the main ones. Could you please confirm this is really needed from @unera or @alyapunov?

Reply:
The implementation provided in #8660 by @alyapunov supports negative indices (merged).